### ISSUE TITLE
Behavior change for `foo and 1 or 2`: 3.12 newly converts `foo` to bool twice

### ISSUE THREAD

**Issue by frigus02** [python/cpython#124285]
# Bug report

### Bug description:

We noticed a behavior change between 3.11 and 3.12. The following code calls `Foo.__bool__` once in 3.11 and twice in 3.12. Consequently for this contrived example, the expression evaluates to different results in 3.11 and 3.12.

```python
class Foo:
    def __init__(self):
        self._a = True
    def __bool__(self):
        self._a = not self._a
        print(f"Foo.__bool__ -> {self._a}")
        return self._a

Foo() and "a string" or 42
```

In Python 3.11:

```
>>> Foo() and "a string" or 42
Foo.__bool__ -> False
42
```

In Python 3.12 (and 3.13.0b2):

```
>>> Foo() and "a string" or 42
Foo.__bool__ -> False
Foo.__bool__ -> True
<__main__.Foo object at 0x7f0ae554c1a0>
```

Is this change intentional?

Note that I'm not necessarily asking to change this. We should arguably change the code to `"a string" if Foo() else 42`, which evaluates the same in 3.11 and 3.12.

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-124394
* gh-124723

**Comment by ZeroIntensity** [comment]
I haven't had a chance to look closer at this, but I will say that flipping the truthiness of something everytime it's used in `__bool__` is quite odd. How is this being used in practice?

**Comment by frigus02** [comment]
We noticed this when upgrading pytype to support 3.12. Pytype evaluates the bytecode and reported that this expression may now now be `Union[Foo, str, int]`. See https://github.com/google/pytype/pull/1777.

After staring at the bytecode for a while we noticed this is really a runtime change and a specifically crafted `__bool__` method could show that. But I don't know of any real example that does that.

Apparently pyright also encodes that behavior somehow. But it does that independently of the Python version. [pyright playground](https://pyright-play.net/?pythonVersion=3.11&code=MYGwhgzhAEBiD28BcAoa7oBMCmAzaA%2BgQJYB2xALkQBQTYi4CUqGr0dDAdAWNALzQAKgCcArtjQYc%2BIgCNEIGhyYs26Zd14DS8Cu3q5NktcOwVRw0vq48UKMEnYVhAHzJ6BCeNUbQwpTGgAIl4IZzIAcyDoeGFoABYAJiA).

**Comment by Eclips4** [comment]
Yeah, this code is quite odd but this is definitely is regression, thanks for the report @frigus02. I'm bisecting the bad commit right know

**Comment by tomasr8** [comment]
Confirmed to be happening on current main as well

**Comment by skirpichev** [comment]
Documentation in 3.12 says: "The expression x and y first evaluates x; if x is false, its value is returned; otherwise, y is evaluated and the resulting value is returned.
The expression x or y first evaluates x; if x is true, its value is returned; otherwise, y is evaluated and the resulting value is returned.
Note that neither [and](https://docs.python.org/3.12/reference/expressions.html#and) nor [or](https://docs.python.org/3.12/reference/expressions.html#or) restrict the value and type they return to False and True, but rather return the last evaluated argument."

That's exactly how it works in 3.12+:
1. Compute Foo() and "a string":
2. Foo() evaluates to False => returns Foo() object
3. Compute Foo() or "42":
4. Foo() evaluated to False => returns Foo() object

```py
# a.py
class Foo:
    def __init__(self):
        self._a = True
    def __bool__(self):
        print(self)
        self._a = not self._a
        print(f"Foo.__bool__ -> {self._a}")
        return self._a
print(Foo() and "a string" or 42)
```

```
$ python3.12 a.py
<__main__.Foo object at 0x7f1b514d6120>
Foo.__bool__ -> False
<__main__.Foo object at 0x7f1b514d6120>
Foo.__bool__ -> True
<__main__.Foo object at 0x7f1b514d6120>
```

It looks, that before the result of evaluation Foo() in (2) was "cached".  That's ok, assuming that bool(obj) would have no side effects.  Which is not our case:
```pycon
>>>
>>> x = Foo()
>>> bool(x)
<a.Foo object at 0x7fb5d7bfbaa0>
Foo.__bool__ -> False
False
>>> bool(x)
<a.Foo object at 0x7fb5d7bfbaa0>
Foo.__bool__ -> True
True
```

@Eclips4, thanks for debugging.  But I don't think it's a bug.  Does documentation says somewhere that ``__bool__()`` should have no side effects?

**Comment by Eclips4** [comment]
It's too late to change something as big as that and say that `__bool__` should have no side effects. In fact, it's essential in Python to expect side effects somewhere.. This is not making `__bool__` an exclusive here. The fact that `__bool__` is called twice is less scary than the fact that the same code has different results on different versions, because the commit which introduced it didn't intend that.

**Comment by skirpichev** [comment]
> the commit which introduced it didn't intend that.

Maybe (let's wait a reply from core dev;)), but IMHO it looks like it fixed a bug (which was mentioned by OP pre-3.12 behaviour).

If we return rules as in 3.11 - we will have to adjust docs accordingly, because old behaviour doesn't match docs (in 3.11 - too).  And yes, ``__bool__()`` - will be a special beast!

**Comment by ZeroIntensity** [comment]
If we do decide to fix this, it's probably going to be too complicated to backport :(

**Comment by ethanfurman** [comment]
It's a bug.  Terms should only be evaluated once.

**Comment by markshannon** [comment]
Thanks @ncoghlan for the analysis.

I think we should classify this as a bug. There is no reason for `bool(a and b)` to be re-evaluated, since we know it from `bool(a)` and `bool(b)`.

@iritkatriel could we fix this by using pseudo instructions similar to the `JUMP_IF_[FALSE|TRUE]_OR_POP` used in 3.11, and converting them after jump-threading?

I note that in https://github.com/python/cpython/commit/ef0a82b68290ddde9a2f5f88bbe02caa195158a5 the instructions back then were `JUMP_IF_[FALSE|TRUE]` not the `JUMP_IF_[FALSE|TRUE]_OR_POP` of 3.11.
We could use the simpler instructions in codegen, as they are straightforward to convert later:
`JUMP_IF_FALSE` -> `COPY 1; TO_BOOL; POP_JUMP_IF_FALSE`.

**Comment by iritkatriel** [comment]
> @iritkatriel could we fix this by using pseudo instructions similar to the JUMP_IF_[FALSE|TRUE]_OR_POP used in 3.11, and converting them after jump-threading?

The pseudo-instruction mechanism currently doesn't support mapping a pseudo-instruction to a sequence of instructions. We could teach the generated to do this, but then we will probably have the problem that this pseudo-instruction has irregular stack effect, which is what we were trying to get away from in the first place. 

So I think probably not with `JUMP_IF_[FALSE|TRUE]_OR_POP`, but maybe with `JUMP_IF_[FALSE|TRUE]`. Let me try to make that work.

**Comment by iritkatriel** [comment]
I merged a fix, but I don't know about back porting it.

**Comment by ncoghlan** [comment]
@JelleZijlstra asked on the 3.14 PR if the idea of doing this in the AST optimiser was worth considering further.

Duplicating the `c` node in the generated AST really isn't desirable if we can get the same performance benefit later in the pipeline without introducing any duplication in the generated code.

I mainly suggested the AST option in case it was hard to restore the optimisation at the opcode generation stage, which doesn't appear to be a relevant concern (since @iritkatriel  already restored it).

**Comment by JelleZijlstra** [comment]
@iritkatriel I think we shouldn't backport this, both because the fix is invasive (new opcodes), and because the behavior change feels too big for a bugfix release.

**Comment by ncoghlan** [comment]
This skip coincidentally came up in a [recent forum thread](https://discuss.python.org/t/implement-do-to-simplify-coding/65076/38).

The point was made that if compilers are allowed to make this assumption (and they are, since the reference interpreter did it for years, and only stopped due to a bug), then the `__bool__` data model entry should mention that.

**Comment by skirpichev** [comment]
Here is an attempt to document things: https://github.com/python/cpython/pull/124723

PS: @Eclips4, probably we should clear 3.12-3.14 labels, as the fix is not going to be backported?
